File: games/tbamud/commands/items.py
"""Item commands — get, drop, put, give, wear, remove, equipment."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from core.engine import Engine
    from core.session import Session
    from core.world import ObjInstance


# Wear position names (tbaMUD standard 18 slots)
WEAR_NAMES = [
    "머리 위에", "왼쪽 손가락에", "오른쪽 손가락에", "목에 (1)",
    "목에 (2)", "몸통에", "머리에", "다리에",
    "발에", "손에", "팔에", "방패로",
    "몸 주위에", "허리에", "왼팔에", "오른팔에",
    "오른손에", "왼손에",
]


def register(engine: Engine) -> None:
    engine.register_command("get", do_get, korean="줍")
    engine.register_command("take", do_get)
    engine.register_command("drop", do_drop, korean="버려")
    engine.register_command("wear", do_wear, korean="입")
    engine.register_command("wield", do_wear)
    engine.register_command("remove", do_remove, korean="벗")
    engine.register_command("equipment", do_equipment, korean="장비")
    engine.register_command("eq", do_equipment)
    engine.register_command("give", do_give, korean="줘")
    engine.register_command("put", do_put, korean="넣")


async def do_get(session: Session, args: str) -> None:
    char = session.character
    if not char:
        return
    if not args:
        await session.send_line("무엇을 주우시겠습니까?")
        return

    room = session.engine.world.get_room(char.room_vnum)
    if not room:
        return

    target = args.strip().lower()

    if target == "all" or target == "모두":
        if not room.objects:
            await session.send_line("여기에는 아무것도 없습니다.")
            return
        picked = list(room.objects)
        for obj in picked:
            room.objects.remove(obj)
            obj.room_vnum = None
            obj.carried_by = char
            char.inventory.append(obj)
            await session.send_line(f"{obj.name}을(를) 주웠습니다.")
        return

    for obj in room.objects:
        if target in obj.proto.keywords.lower():
            room.objects.remove(obj)
            obj.room_vnum = None
            obj.carried_by = char
            char.inventory.append(obj)
            await session.send_line(f"{obj.name}을(를) 주웠습니다.")
            return

    await session.send_line("그런 것은 여기에 없습니다.")


async def do_drop(session: Session, args: str) -> None:
    char = session.character
    if not char:
        return
    if not args:
        await session.send_line("무엇을 버리시겠습니까?")
        return

    target = args.strip().lower()

    for obj in char.inventory:
        if target in obj.proto.keywords.lower():
            char.inventory.remove(obj)
            obj.carried_by = None
            session.engine.world.obj_to_room(obj, char.room_vnum)
            await session.send_line(f"{obj.name}을(를) 버렸습니다.")
            return

    await session.send_line("그런 것을 갖고 있지 않습니다.")


async def do_wear(session: Session, args: str) -> None:
    char = session.character
    if not char:
        return
    if not args:
        await session.send_line("무엇을 착용하시겠습니까?")
        return

    target = args.strip().lower()

    for obj in char.inventory:
        if target in obj.proto.keywords.lower():
            # Find first available wear position from obj's wear_flags
            wear_flags = obj.proto.wear_flags
            if not wear_flags or wear_flags == [0]:
                await session.send_line("그것은 착용할 수 없습니다.")
                return
            # Simple: use first wear flag > 0 as position
            pos = -1
            for flag in wear_flags:
                if flag > 0 and flag < len(WEAR_NAMES):
                    if flag not in char.equipment:
                        pos = flag
                        break
            if pos < 0:
                await session.send_line("이미 그 위치에 뭔가를 착용하고 있습니다.")
                return
            char.inventory.remove(obj)
            obj.carried_by = None
            obj.worn_by = char
            obj.wear_pos = pos
            char.equipment[pos] = obj
            pos_name = WEAR_NAMES[pos] if pos < len(WEAR_NAMES) else "어딘가에"
            await session.send_line(f"{obj.name}을(를) {pos_name} 착용했습니다.")
            return

    await session.send_line("그런 것을 갖고 있지 않습니다.")


async def do_remove(session: Session, args: str) -> None:
    char = session.character
    if not char:
        return
    if not args:
        await session.send_line("무엇을 벗으시겠습니까?")
        return

    target = args.strip().lower()

    for pos, obj in list(char.equipment.items()):
        if target in obj.proto.keywords.lower():
            del char.equipment[pos]
            obj.worn_by = None
            obj.wear_pos = -1
            obj.carried_by = char
            char.inventory.append(obj)
            await session.send_line(f"{obj.name}을(를) 벗었습니다.")
            return

    await session.send_line("그런 것을 착용하고 있지 않습니다.")


async def do_equipment(session: Session, args: str) -> None:
    char = session.character
    if not char:
        return
    if not char.equipment:
        await session.send_line("아무것도 착용하고 있지 않습니다.")
        return
    await session.send_line("착용 중인 장비:")
    for pos in sorted(char.equipment.keys()):
        obj = char.equipment[pos]
        pos_name = WEAR_NAMES[pos] if pos < len(WEAR_NAMES) else "어딘가"
        await session.send_line(f"  <{pos_name}> {obj.name}")


async def do_give(session: Session, args: str) -> None:
    if not args:
        await session.send_line("무엇을 누구에게 주시겠습니까?")
        return
    await session.send_line("구현 예정입니다.")


async def do_put(session: Session, args: str) -> None:
    if not args:
        await session.send_line("무엇을 어디에 넣으시겠습니까?")
        return
    await session.send_line("구현 예정입니다.")

File: games/tbamud/lua/commands/items.lua
-- items.lua — tbaMUD item commands (overrides common get/drop, adds wear/remove)

-- Wear position names (tbaMUD 18 slots)
local WEAR_NAMES = {
    [0]  = "머리 위에",
    [1]  = "왼쪽 손가락에",
    [2]  = "오른쪽 손가락에",
    [3]  = "목에 (1)",
    [4]  = "목에 (2)",
    [5]  = "몸통에",
    [6]  = "머리에",
    [7]  = "다리에",
    [8]  = "발에",
    [9]  = "손에",
    [10] = "팔에",
    [11] = "방패로",
    [12] = "몸 주위에",
    [13] = "허리에",
    [14] = "왼팔에",
    [15] = "오른팔에",
    [16] = "오른손에",
    [17] = "왼손에",
}

-- ── get/take (override common — adds "all" support) ─────────────

register_command("get", function(ctx, args)
    local ch = ctx.char
    if not ch then return end
    if not args or args == "" then
        ctx:send("무엇을 주우시겠습니까?")
        return
    end

    local target = args:lower()
    local room = ctx:get_room()
    if not room then return end
    local objects = ctx:get_objects()

    if target == "all" or target == "모두" then
        if #objects == 0 then
            ctx:send("여기에는 아무것도 없습니다.")
            return
        end
        for i = 1, #objects do
            local obj = objects[i]
            ctx:obj_from_room(obj)
            ctx:obj_to_char(obj, ch)
            ctx:send(obj.name .. "을(를) 주웠습니다.")
        end
        return
    end

    for i = 1, #objects do
        local obj = objects[i]
        if obj.proto.keywords:lower():find(target, 1, true) then
            ctx:obj_from_room(obj)
            ctx:obj_to_char(obj, ch)
            ctx:send(obj.name .. "을(를) 주웠습니다.")
            return
        end
    end
    ctx:send("그런 것은 여기에 없습니다.")
end, "줍")

register_command("take", function(ctx, args)
    ctx:call_command("get", args or "")
end)

-- ── drop ────────────────────────────────────────────────────────

register_command("drop", function(ctx, args)
    local ch = ctx.char
    if not ch then return end
    if not args or args == "" then
        ctx:send("무엇을 버리시겠습니까?")
        return
    end

    local obj = ctx:find_obj_inv(args)
    if not obj then
        ctx:send("그런 것을 갖고 있지 않습니다.")
        return
    end
    ctx:obj_from_char(obj)
    ctx:obj_to_room(obj, ch.room_vnum)
    ctx:send(obj.name .. "을(를) 버렸습니다.")
end, "버려")

-- ── wear/wield ──────────────────────────────────────────────────

register_command("wear", function(ctx, args)
    local ch = ctx.char
    if not ch then return end
    if not args or args == "" then
        ctx:send("무엇을 착용하시겠습니까?")
        return
    end

    local obj = ctx:find_obj_inv(args)
    if not obj then
        ctx:send("그런 것을 갖고 있지 않습니다.")
        return
    end

    local wear_flags = obj.proto.wear_flags
    if not wear_flags then
        ctx:send("그것은 착용할 수 없습니다.")
        return
    end

    -- Find first available wear position > 0
    local pos = -1
    local ok, len = pcall(function() return #wear_flags end)
    if ok then
        for i = 0, len - 1 do
            local fok, flag = pcall(function() return wear_flags[i] end)
            if fok and flag and flag > 0 and flag < 18 then
                local equip = ctx:get_equipment()
                local occupied = false
                for j = 1, #equip do
                    if equip[j].slot == flag then
                        occupied = true
                        break
                    end
                end
                if not occupied then
                    pos = flag
                    break
                end
            end
        end
    end

    if pos < 0 then
        ctx:send("이미 그 위치에 뭔가를 착용하고 있습니다.")
        return
    end

    ctx:obj_from_char(obj)
    ctx:equip(obj, pos)
    local pos_name = WEAR_NAMES[pos] or "어딘가에"
    ctx:send(obj.name .. "을(를) " .. pos_name .. " 착용했습니다.")
end, "입")

register_command("wield", function(ctx, args)
    ctx:call_command("wear", args or "")
end)

-- ── remove ──────────────────────────────────────────────────────

register_command("remove", function(ctx, args)
    local ch = ctx.char
    if not ch then return end
    if not args or args == "" then
        ctx:send("무엇을 벗으시겠습니까?")
        return
    end

    local target = args:lower()
    local equip = ctx:get_equipment()
    for i = 1, #equip do
        local obj = equip[i].obj
        if obj.proto.keywords:lower():find(target, 1, true) then
            ctx:unequip(equip[i].slot)
            ctx:send(obj.name .. "을(를) 벗었습니다.")
            return
        end
    end
    ctx:send("그런 것을 착용하고 있지 않습니다.")
end, "벗")

-- ── give ────────────────────────────────────────────────────────

register_command("give", function(ctx, args)
    if not args or args == "" then
        ctx:send("무엇을 누구에게 주시겠습니까?")
        return
    end
    ctx:send("구현 예정입니다.")
end, "줘")

-- ── put ─────────────────────────────────────────────────────────

register_command("put", function(ctx, args)
    if not args or args == "" then
        ctx:send("무엇을 어디에 넣으시겠습니까?")
        return
    end
    ctx:send("구현 예정입니다.")
end, "넣")
